riscv_executor/src/lib.rs

```diff
@@ -0,0 +1,784 @@
+        Self(0)
+    }
+
+    fn u(&self) -> u32 {
+        self.0.try_into().unwrap()
+    }
+
+    fn s(&self) -> i32 {
+        self.0.try_into().unwrap()
+    }
+}
+
+impl From<u32> for Elem {
+    fn from(value: u32) -> Self {
+        Self(value as i64)
+    }
+}
+
+impl From<i64> for Elem {
+    fn from(value: i64) -> Self {
+        Self(value)
+    }
+}
+
+impl From<i32> for Elem {
+    fn from(value: i32) -> Self {
+        Self(value as i64)
+    }
+}
+
+pub struct ExecutionTrace<'a> {
+    reg_map: HashMap<&'a str, usize>,
+
+    /// Values of the registers in the execution trace.
+    ///
+    /// Each N elements is a row with all registers.
+    values: Vec<Elem>,
+}
+
+mod builder {
+    use std::collections::HashMap;
+
+    use ast::asm_analysis::Machine;
+    use number::FieldElement;
+
+    use crate::{Elem, ExecutionTrace};
+
+    fn register_names<T: FieldElement>(main: &Machine<T>) -> Vec<&str> {
+        main.registers.iter().map(|stmnt| &stmnt.name[..]).collect()
+    }
+
+    pub struct TraceBuilder<'a, 'b> {
+        trace: ExecutionTrace<'a>,
+
+        /// First register of current row.
+        /// Next row is reg_map.len() elems ahead.
+        curr_idx: usize,
+
+        // index of special case registers to look after:
+        x0_idx: usize,
+        pc_idx: usize,
+
+        /// The PC in the register bank refers to the batches, we have to track our
+        /// actual program counter independently.
+        next_statement_line: u32,
+
+        /// When PC is written, we need to know what line to actually execute next
+        /// from this map of batch to statement line.
+        batch_to_line_map: &'b [u32],
+    }
+
+    impl<'a, 'b> TraceBuilder<'a, 'b> {
+        // creates a new builder
+        pub fn new<T: FieldElement>(main: &'a Machine<T>, batch_to_line_map: &'b [u32]) -> Self {
+            let reg_map = register_names(main)
+                .into_iter()
+                .enumerate()
+                .map(|(i, name)| (name, i))
+                .collect::<HashMap<&str, usize>>();
+
+            // first row has all values zeroed
+            let values = vec![Elem::zero(); 2 * reg_map.len()];
+
+            let mut ret = Self {
+                curr_idx: 0,
+                x0_idx: reg_map["x0"],
+                pc_idx: reg_map["pc"],
+                trace: ExecutionTrace { reg_map, values },
+                next_statement_line: 1,
+                batch_to_line_map,
+            };
+
+            ret.set_next_pc();
+
+            ret
+        }
+
+        /// get current value of register
+        pub(crate) fn g(&self, idx: &str) -> Elem {
+            self.g_idx(self.trace.reg_map[idx])
+        }
+
+        /// get current value of register by register index instead of name
+        fn g_idx(&self, idx: usize) -> Elem {
+            self.trace.values[self.curr_idx + idx]
+        }
+
+        /// set next value of register
+        pub(crate) fn s(&mut self, idx: &str, value: impl Into<Elem>) {
```

Comment:
Wouldn't it be much better readable if `s` would return a mutable reference to the value?

Reply:
In my subjective opinion, no.

Instead of:
```rs
obj.s("pc", x + 42);
```
it would be
```rs
*obj.s("pc") = (x + 42).into();
```
or more conventionally
```rs
*obj.g_mut("pc") = (x + 42).into();
```

But most importantly, I can't have the logic to handle PC or x0 if I return a mut reference.